example/09_convnd_fwd/convnd_fwd_common.hpp

```diff
@@ -27,6 +27,88 @@ void print_helper_msg()
               << ck::utils::conv::get_conv_param_parser_helper_msg() << std::endl;
 }
 
+template <typename DataType>
+inline __host__ __device__ constexpr double get_rtol()
+{
+    if constexpr(std::is_same_v<DataType, float>)
```

Comment:
Can We change it to more generic? I propose to calculate it in this way:
`C * Z * Y * X * 2^-mantisa`
For example for 3 multiple adds:
```
(((x1*x2)*(1+eps) + (x3*x4))(1+eps) + (x5+6)(1+eps))(1+eps)= 
((x1 *x2 + x1*x2 * eps + x3 * x4 + x3*x4*eps) (1+eps)) + x5*x6 + x5*x6*eps)(1+eps) = 
(x1 *x2 + x1*x2 * eps + x3 * x4 + x3*x4*eps + x1 *x2  * eps+ x1*x2 * eps * eps + x3 * x4 * eps + x3*x4*eps * eps  + x5*x6+ x5*x6*eps)(1+eps)
```
We can skip eps*eps, it is very very low, then:
```
(x1 *x2 + x1*x2 * eps + x3 * x4 + x3*x4*eps + x1 *x2  * eps+  x3 * x4 * eps +  x5*x6+ x5*x6*eps)(1+eps) = 
x1 *x2 + x1*x2 * eps + x3 * x4 + x3*x4*eps + x1 *x2  * eps+  x3 * x4 * eps +  x5*x6+ x5*x6*eps + x1 *x2 * eps + x1*x2 * eps * eps + x3 * x4 + x3*x4*eps * eps + x1 *x2  * eps * eps+  x3 * x4 * eps * eps +  x5*x6+ x5*x6*eps * eps
```
Skip eps* eps again:
```
x1 *x2 + x1*x2 * eps + x3 * x4 + x3*x4*eps + x1 *x2  * eps+  x3 * x4 * eps +  x5*x6+ x5*x6*eps + x1 *x2 * eps = 
x1 +x2  +x3*x4 +x5*x6 + 3*x1*x2*eps + 2 x3*x4 eps + x5 *x6*eps
```
So for 3 multiple + add maximum  it should not exceed `3*x1*x2*eps + 2 x3*x4 eps + x5 *x6*eps` which should not exceed `3 * 2^-mantisa`.
For each convolution output cell we have `C * Z * Y * X` multiple+add.

Reply:
Hi @bartekxk, thanks for the detailed explanation! I like the idea to make acceptable tolerances meaningful, but I would like to clarify couple more details:
1. The logic you are suggesting is for relative tolerance calculation. We could also calculate acceptable absolute tolerance for various datatypes, do you think it is worth it?
2. Problem dimensions are runtime arguments, so we cannot calculate tolerances in compile-time then. I don't think the runtime overhead can become significant, and it is not kernel-related, but there will be some change.
3. From the low precision type perspective, there are several 8-bit float standards we are going to support, so calculating exact tolerances would make maintenance harder. For example, for various architectures we will have to add macros and calculate corresponding tolerances.
Looking forward to hear your thoughts! Overall, I'm totally for updating our tolerance logic repo-wide and I believe the only way to do it is one operator at a time ;-)

Reply:
1. Yes, this is for relative, the same example for absolute:
`((((x1*x2 + eps)  + (x3*x4 + eps)) + eps) + (x5*x6 +eps)) + eps = 
((x1*x2 + x3 * x4 + 3* eps) + (x5*x6 + eps) )+eps  =
x1 *x2 + x3 * x4 + x5*x6 + 5 * eps`
So we can assume that it is `2 * multiple add operations * 0.5 * ULP =  2 * C * Z * Y * X * 0.5 * ULP`, It is a little bit harder to specify ULP for each temporary value during calculation. One idea is to use maximum value from GenerateTensorValue. We can discuss it. 
2. Yes, but in this case we can just assume relative error equal to `C * Z * Y * X * 2^-mantisa`. Calculation of this formula should be very fast.
3. Maybe we can create some custom function like `ck::epsilon<DataType>()` to return 2^-mantisa for each DataType.

This change could be also done in seperate change which fix this calculation in whole repository :). I will be happy to help you implement this.

Reply:
We agreed to implement such threshold calculation in next PR.